apps/src/codegen.js

```diff
@@ -48,6 +48,31 @@ exports.evalWith = function (code, options, legacy) {
   }
 };
 
+exports.evalWithEvents = function (apis, events, generator) {
+  let interpreter, currentCallback;
+  const hooks = {};
+  const code = Object.keys(events).map(event => {
+    hooks[event] = () => {
+      currentCallback(event);
+      interpreter.run();
+    };
+    return `function ${event} () {
+      ${generator(events[event])}
+    }
+    `;
+  }).join('');
```

Comment:
I might split this into two different methods rather than mutating hooks in the middle of mapping, i.e. first do a `forEach` that sets hooks, then do a `map` that creates our code.

You could potentially even replace the first forEach with `_.fromPairs` i.e.
```
const hooks = _.fromPairs(Object.keys(events).map(event => [
  event,
  () => {
    currentCallback(event);
    interpreter.run()
  }
])
```
Not sure if/how much clearer that ends up being.